[ML-Dataframe] add basic configuration

hendrikmuhs opened the pull request:
**FEATURE BRANCH PR**

add basic configuration for feature index builder job, reads the source, destination index, aggregation source and aggregation config

Note: 

The AggregationConfig awaits a fix upstream, see https://github.com/elastic/elasticsearch/pull/33942

elasticmachine commented:
Pinging @elastic/ml-core

droberts195 commented on x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/FeatureIndexBuilderJobConfig.java:
```diff
@@ -28,25 +30,55 @@
+    private final SourceConfig sourceConfig;
+    private final AggregationConfig aggregationConfig;
 
     private static final ConstructingObjectParser<FeatureIndexBuilderJobConfig, String> PARSER = new ConstructingObjectParser<>(NAME, false,
```
Experience has shown that for stored configs it's best to have both a lenient parser that ignores unknown fields and a strict parser that throws an error on encountering an unknown field.  See for example `AnalysisConfig.LENIENT_PARSER` and `AnalysisConfig.STRICT_PARSER`.  Then use the strict parser when parsing user-supplied requests and the lenient parser when parsing configs retrieved from storage.  That way things don't completely break in a mixed version cluster when the newer version has added an extra field to the config.

hendrikmuhs replied:
made a note int the meta issue (https://github.com/elastic/elasticsearch/issues/33600), as we will highly likely move out of the cluster state (https://github.com/elastic/elasticsearch/issues/33952) the config needs a full revisit

droberts195 commented on x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/SourceConfig.java:
```diff
@@ -0,0 +1,110 @@
+        this.sources = new ArrayList<>(num);
+        for (int i = 0; i < num; i++) {
+            CompositeValuesSourceBuilder<?> builder = CompositeValuesSourceParserHelper.readFrom(in);
+            getSources().add(builder);
```
It would be nice if this class was immutable, and this line shows why it isn't currently.  This is how to make it immutable:

```
        List<CompositeValuesSourceBuilder<?>> sources = new ArrayList<>(num);
        for (int i = 0; i < num; i++) {
            CompositeValuesSourceBuilder<?> builder = CompositeValuesSourceParserHelper.readFrom(in);
            sources.add(builder);
        }
        this.sources = Collections.unmodifableList(sources);
```

droberts195 commented on x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/SourceConfig.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    public SourceConfig(List<CompositeValuesSourceBuilder<?>> sources) {
+        this.sources = sources;
```
For immutability:

```
    this.sources = Collections.unmodifiableList(new ArrayList<>(sources));
```

droberts195 commented on x-pack/plugin/ml-feature-index-builder/src/test/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/AbstractSerializingFeatureIndexBuilderTestCase.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
```
It's unusual to override the base class `@Before` method and then call the base class version first.  JUnit will by default call all the base class `@Before` methods before the derived class `@Before` methods, so it would be more usual to just give the derived class `@Before` method a different name and let JUnit call the two methods in the standard order, avoiding the need to call `super.setUp()` here.

droberts195 commented on x-pack/plugin/ml-feature-index-builder/src/test/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/AggregationConfigTests.java:
```diff
@@ -0,0 +1,65 @@
+    public static AggregationConfig randonAggregationConfig() {
+        AggregatorFactories.Builder builder = new AggregatorFactories.Builder();
+
+        for (int i = 1; i < randomIntBetween(1, 20); ++i) {
```
Is it valid for the number of aggregator factories to be 0?  If not maybe the `AggregationConfig` class itself should validate that it has at least 1?

hendrikmuhs replied:
good catch, should be `for (int i = 0 ...`

Nevertheless, `AggregationConfig` misses validation, will be part of an upcoming PR.

droberts195 commented on x-pack/plugin/ml-feature-index-builder/src/test/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/SourceConfigTests.java:
```diff
@@ -0,0 +1,59 @@
+public class SourceConfigTests extends AbstractSerializingFeatureIndexBuilderTestCase<SourceConfig> {
+
+    public static SourceConfig randomSourceConfig() {
+        int numSources = randomIntBetween(1, 10);
```
Is it valid for the number of sources to be 0? If not maybe the `SourceConfig` class itself should validate that it has at least 1?

hendrikmuhs replied:
the for-loop below counts from `0`, so there will be at least 1 source for this test.

`SourceConfig` lacks validation at the moment, to be addressed in an upcoming PR.

droberts195 commented on x-pack/plugin/ml-feature-index-builder/src/test/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/AggregationConfigTests.java:
```diff
@@ -0,0 +1,65 @@
+@AwaitsFix(bugUrl="https://github.com/elastic/elasticsearch/issues/33942")
+public class AggregationConfigTests extends AbstractSerializingFeatureIndexBuilderTestCase<AggregationConfig> {
+
+    public static AggregationConfig randonAggregationConfig() {
```
typo: randon -> random

droberts195 commented on x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/FeatureIndexBuilderJobConfig.java:
```diff
@@ -28,25 +30,55 @@
+        indexPattern = in.readString();
+        destinationIndex = in.readString();
+        sourceConfig = in.readOptionalWriteable(SourceConfig::new);
+        aggregationConfig = in.readOptionalWriteable(AggregationConfig::new);
```
It seems like there are unnecessarily many levels where `null` is allowed.  You're allowing `aggregationConfig` to be `null` here, but also in `AggregationConfig` `aggregatorFactoryBuilder` is allowed to be `null`.  I think at most one of these possibilities should be allowed.

hendrikmuhs replied:
fixed in `AggregationConfig` - not aiming for at the moment but in future I see the possibility of this builder to not use aggregations.

droberts195 commented on x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/AggregationConfig.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    AggregationConfig(final StreamInput in) throws IOException {
+        aggregatorFactoryBuilder = in.readOptionalWriteable(AggregatorFactories.Builder::new);
```
It seems like there are unnecessarily many levels where `null` is allowed.  You're allowing `aggregatorFactoryBuilder` to be `null` here, but also in `FeatureIndexBuilderJobConfig` `aggregationConfig` is allowed to be `null`.  I think at most one of these possibilities should be allowed.

hendrikmuhs commented:
@droberts195 Thanks for the review, addressed most of your comments. 

For the rest I am planning to tackle them in separate PR's - so many open tasks at the moment.
